<!doctype html>
<html>
<head>
<meta charset='UTF-8'><meta name='viewport' content='width=device-width initial-scale=1'>
<title>README</title><link href='https://fonts.loli.net/css?family=Open+Sans:400italic,700italic,700,400&subset=latin,latin-ext' rel='stylesheet' type='text/css' /><style type='text/css'>html {overflow-x: initial !important;}:root { --bg-color:#ffffff; --text-color:#333333; --select-text-bg-color:#B5D6FC; --select-text-font-color:auto; --monospace:"Lucida Console",Consolas,"Courier",monospace; }
html { font-size: 14px; background-color: var(--bg-color); color: var(--text-color); font-family: "Helvetica Neue", Helvetica, Arial, sans-serif; -webkit-font-smoothing: antialiased; }
body { margin: 0px; padding: 0px; height: auto; bottom: 0px; top: 0px; left: 0px; right: 0px; font-size: 1rem; line-height: 1.42857; overflow-x: hidden; background: inherit; tab-size: 4; }
iframe { margin: auto; }
a.url { word-break: break-all; }
a:active, a:hover { outline: 0px; }
.in-text-selection, ::selection { text-shadow: none; background: var(--select-text-bg-color); color: var(--select-text-font-color); }
#write { margin: 0px auto; height: auto; width: inherit; word-break: normal; overflow-wrap: break-word; position: relative; white-space: normal; overflow-x: visible; padding-top: 40px; }
#write.first-line-indent p { text-indent: 2em; }
#write.first-line-indent li p, #write.first-line-indent p * { text-indent: 0px; }
#write.first-line-indent li { margin-left: 2em; }
.for-image #write { padding-left: 8px; padding-right: 8px; }
body.typora-export { padding-left: 30px; padding-right: 30px; }
.typora-export .footnote-line, .typora-export li, .typora-export p { white-space: pre-wrap; }
@media screen and (max-width: 500px) {
  body.typora-export { padding-left: 0px; padding-right: 0px; }
  #write { padding-left: 20px; padding-right: 20px; }
  .CodeMirror-sizer { margin-left: 0px !important; }
  .CodeMirror-gutters { display: none !important; }
}
#write li > figure:last-child { margin-bottom: 0.5rem; }
#write ol, #write ul { position: relative; }
img { max-width: 100%; vertical-align: middle; image-orientation: from-image; }
button, input, select, textarea { color: inherit; font: inherit; }
input[type="checkbox"], input[type="radio"] { line-height: normal; padding: 0px; }
*, ::after, ::before { box-sizing: border-box; }
#write h1, #write h2, #write h3, #write h4, #write h5, #write h6, #write p, #write pre { width: inherit; }
#write h1, #write h2, #write h3, #write h4, #write h5, #write h6, #write p { position: relative; }
p { line-height: inherit; }
h1, h2, h3, h4, h5, h6 { break-after: avoid-page; break-inside: avoid; orphans: 4; }
p { orphans: 4; }
h1 { font-size: 2rem; }
h2 { font-size: 1.8rem; }
h3 { font-size: 1.6rem; }
h4 { font-size: 1.4rem; }
h5 { font-size: 1.2rem; }
h6 { font-size: 1rem; }
.md-math-block, .md-rawblock, h1, h2, h3, h4, h5, h6, p { margin-top: 1rem; margin-bottom: 1rem; }
.hidden { display: none; }
.md-blockmeta { color: rgb(204, 204, 204); font-weight: 700; font-style: italic; }
a { cursor: pointer; }
sup.md-footnote { padding: 2px 4px; background-color: rgba(238, 238, 238, 0.7); color: rgb(85, 85, 85); border-radius: 4px; cursor: pointer; }
sup.md-footnote a, sup.md-footnote a:hover { color: inherit; text-transform: inherit; text-decoration: inherit; }
#write input[type="checkbox"] { cursor: pointer; width: inherit; height: inherit; }
figure { overflow-x: auto; margin: 1.2em 0px; max-width: calc(100% + 16px); padding: 0px; }
figure > table { margin: 0px; }
tr { break-inside: avoid; break-after: auto; }
thead { display: table-header-group; }
table { border-collapse: collapse; border-spacing: 0px; width: 100%; overflow: auto; break-inside: auto; text-align: left; }
table.md-table td { min-width: 32px; }
.CodeMirror-gutters { border-right: 0px; background-color: inherit; }
.CodeMirror-linenumber { user-select: none; }
.CodeMirror { text-align: left; }
.CodeMirror-placeholder { opacity: 0.3; }
.CodeMirror pre { padding: 0px 4px; }
.CodeMirror-lines { padding: 0px; }
div.hr:focus { cursor: none; }
#write pre { white-space: pre-wrap; }
#write.fences-no-line-wrapping pre { white-space: pre; }
#write pre.ty-contain-cm { white-space: normal; }
.CodeMirror-gutters { margin-right: 4px; }
.md-fences { font-size: 0.9rem; display: block; break-inside: avoid; text-align: left; overflow: visible; white-space: pre; background: inherit; position: relative !important; }
.md-diagram-panel { width: 100%; margin-top: 10px; text-align: center; padding-top: 0px; padding-bottom: 8px; overflow-x: auto; }
#write .md-fences.mock-cm { white-space: pre-wrap; }
.md-fences.md-fences-with-lineno { padding-left: 0px; }
#write.fences-no-line-wrapping .md-fences.mock-cm { white-space: pre; overflow-x: auto; }
.md-fences.mock-cm.md-fences-with-lineno { padding-left: 8px; }
.CodeMirror-line, twitterwidget { break-inside: avoid; }
.footnotes { opacity: 0.8; font-size: 0.9rem; margin-top: 1em; margin-bottom: 1em; }
.footnotes + .footnotes { margin-top: 0px; }
.md-reset { margin: 0px; padding: 0px; border: 0px; outline: 0px; vertical-align: top; background: 0px 0px; text-decoration: none; text-shadow: none; float: none; position: static; width: auto; height: auto; white-space: nowrap; cursor: inherit; -webkit-tap-highlight-color: transparent; line-height: normal; font-weight: 400; text-align: left; box-sizing: content-box; direction: ltr; }
li div { padding-top: 0px; }
blockquote { margin: 1rem 0px; }
li .mathjax-block, li p { margin: 0.5rem 0px; }
li { margin: 0px; position: relative; }
blockquote > :last-child { margin-bottom: 0px; }
blockquote > :first-child, li > :first-child { margin-top: 0px; }
.footnotes-area { color: rgb(136, 136, 136); margin-top: 0.714rem; padding-bottom: 0.143rem; white-space: normal; }
#write .footnote-line { white-space: pre-wrap; }
@media print {
  body, html { border: 1px solid transparent; height: 99%; break-after: avoid; break-before: avoid; font-variant-ligatures: no-common-ligatures; }
  #write { margin-top: 0px; padding-top: 0px; border-color: transparent !important; }
  .typora-export * { -webkit-print-color-adjust: exact; }
  html.blink-to-pdf { font-size: 13px; }
  .typora-export #write { padding-left: 32px; padding-right: 32px; padding-bottom: 0px; break-after: avoid; }
  .typora-export #write::after { height: 0px; }
  .is-mac table { break-inside: avoid; }
}
.footnote-line { margin-top: 0.714em; font-size: 0.7em; }
a img, img a { cursor: pointer; }
pre.md-meta-block { font-size: 0.8rem; min-height: 0.8rem; white-space: pre-wrap; background: rgb(204, 204, 204); display: block; overflow-x: hidden; }
p > .md-image:only-child:not(.md-img-error) img, p > img:only-child { display: block; margin: auto; }
#write.first-line-indent p > .md-image:only-child:not(.md-img-error) img { left: -2em; position: relative; }
p > .md-image:only-child { display: inline-block; width: 100%; }
#write .MathJax_Display { margin: 0.8em 0px 0px; }
.md-math-block { width: 100%; }
.md-math-block:not(:empty)::after { display: none; }
[contenteditable="true"]:active, [contenteditable="true"]:focus, [contenteditable="false"]:active, [contenteditable="false"]:focus { outline: 0px; box-shadow: none; }
.md-task-list-item { position: relative; list-style-type: none; }
.task-list-item.md-task-list-item { padding-left: 0px; }
.md-task-list-item > input { position: absolute; top: 0px; left: 0px; margin-left: -1.2em; margin-top: calc(1em - 10px); border: none; }
.math { font-size: 1rem; }
.md-toc { min-height: 3.58rem; position: relative; font-size: 0.9rem; border-radius: 10px; }
.md-toc-content { position: relative; margin-left: 0px; }
.md-toc-content::after, .md-toc::after { display: none; }
.md-toc-item { display: block; color: rgb(65, 131, 196); }
.md-toc-item a { text-decoration: none; }
.md-toc-inner:hover { text-decoration: underline; }
.md-toc-inner { display: inline-block; cursor: pointer; }
.md-toc-h1 .md-toc-inner { margin-left: 0px; font-weight: 700; }
.md-toc-h2 .md-toc-inner { margin-left: 2em; }
.md-toc-h3 .md-toc-inner { margin-left: 4em; }
.md-toc-h4 .md-toc-inner { margin-left: 6em; }
.md-toc-h5 .md-toc-inner { margin-left: 8em; }
.md-toc-h6 .md-toc-inner { margin-left: 10em; }
@media screen and (max-width: 48em) {
  .md-toc-h3 .md-toc-inner { margin-left: 3.5em; }
  .md-toc-h4 .md-toc-inner { margin-left: 5em; }
  .md-toc-h5 .md-toc-inner { margin-left: 6.5em; }
  .md-toc-h6 .md-toc-inner { margin-left: 8em; }
}
a.md-toc-inner { font-size: inherit; font-style: inherit; font-weight: inherit; line-height: inherit; }
.footnote-line a:not(.reversefootnote) { color: inherit; }
.md-attr { display: none; }
.md-fn-count::after { content: "."; }
code, pre, samp, tt { font-family: var(--monospace); }
kbd { margin: 0px 0.1em; padding: 0.1em 0.6em; font-size: 0.8em; color: rgb(36, 39, 41); background: rgb(255, 255, 255); border: 1px solid rgb(173, 179, 185); border-radius: 3px; box-shadow: rgba(12, 13, 14, 0.2) 0px 1px 0px, rgb(255, 255, 255) 0px 0px 0px 2px inset; white-space: nowrap; vertical-align: middle; }
.md-comment { color: rgb(162, 127, 3); opacity: 0.8; font-family: var(--monospace); }
code { text-align: left; vertical-align: initial; }
a.md-print-anchor { white-space: pre !important; border-width: initial !important; border-style: none !important; border-color: initial !important; display: inline-block !important; position: absolute !important; width: 1px !important; right: 0px !important; outline: 0px !important; background: 0px 0px !important; text-decoration: initial !important; text-shadow: initial !important; }
.md-inline-math .MathJax_SVG .noError { display: none !important; }
.html-for-mac .inline-math-svg .MathJax_SVG { vertical-align: 0.2px; }
.md-math-block .MathJax_SVG_Display { text-align: center; margin: 0px; position: relative; text-indent: 0px; max-width: none; max-height: none; min-height: 0px; min-width: 100%; width: auto; overflow-y: hidden; display: block !important; }
.MathJax_SVG_Display, .md-inline-math .MathJax_SVG_Display { width: auto; margin: inherit; display: inline-block !important; }
.MathJax_SVG .MJX-monospace { font-family: var(--monospace); }
.MathJax_SVG .MJX-sans-serif { font-family: sans-serif; }
.MathJax_SVG { display: inline; font-style: normal; font-weight: 400; line-height: normal; zoom: 90%; text-indent: 0px; text-align: left; text-transform: none; letter-spacing: normal; word-spacing: normal; overflow-wrap: normal; white-space: nowrap; float: none; direction: ltr; max-width: none; max-height: none; min-width: 0px; min-height: 0px; border: 0px; padding: 0px; margin: 0px; }
.MathJax_SVG * { transition: none 0s ease 0s; }
.MathJax_SVG_Display svg { vertical-align: middle !important; margin-bottom: 0px !important; margin-top: 0px !important; }
.os-windows.monocolor-emoji .md-emoji { font-family: "Segoe UI Symbol", sans-serif; }
.md-diagram-panel > svg { max-width: 100%; }
[lang="flow"] svg, [lang="mermaid"] svg { max-width: 100%; height: auto; }
[lang="mermaid"] .node text { font-size: 1rem; }
table tr th { border-bottom: 0px; }
video { max-width: 100%; display: block; margin: 0px auto; }
iframe { max-width: 100%; width: 100%; border: none; }
.highlight td, .highlight tr { border: 0px; }
svg[id^="mermaidChart"] { line-height: 1em; }
mark { background: rgb(255, 255, 0); color: rgb(0, 0, 0); }
.md-html-inline .md-plain, .md-html-inline strong, mark .md-inline-math, mark strong { color: inherit; }
mark .md-meta { color: rgb(0, 0, 0); opacity: 0.3 !important; }


:root {
    --side-bar-bg-color: #fafafa;
    --control-text-color: #777;
}

@include-when-export url(https://fonts.loli.net/css?family=Open+Sans:400italic,700italic,700,400&subset=latin,latin-ext);

html {
    font-size: 16px;
}

body {
    font-family: "Noto Sans CJK TC","Open Sans","Clear Sans","Helvetica Neue",Helvetica,Arial,sans-serif;
    color: rgb(51, 51, 51);
    line-height: 1.6;
}

#write {
    max-width: 860px;
  	margin: 0 auto;
  	padding: 30px;
    padding-bottom: 100px;
}

@media only screen and (min-width: 1400px) {
	#write {
		max-width: 1024px;
	}
}

@media only screen and (min-width: 1800px) {
	#write {
		max-width: 1200px;
	}
}

#write > ul:first-child,
#write > ol:first-child{
    margin-top: 30px;
}

a {
    color: #4183C4;
}
h1,
h2,
h3,
h4,
h5,
h6 {
    position: relative;
    margin-top: 1rem;
    margin-bottom: 1rem;
    font-weight: bold;
    line-height: 1.4;
    cursor: text;
}
h1:hover a.anchor,
h2:hover a.anchor,
h3:hover a.anchor,
h4:hover a.anchor,
h5:hover a.anchor,
h6:hover a.anchor {
    text-decoration: none;
}
h1 tt,
h1 code {
    font-size: inherit;
}
h2 tt,
h2 code {
    font-size: inherit;
}
h3 tt,
h3 code {
    font-size: inherit;
}
h4 tt,
h4 code {
    font-size: inherit;
}
h5 tt,
h5 code {
    font-size: inherit;
}
h6 tt,
h6 code {
    font-size: inherit;
}
h1 {
    padding-bottom: .3em;
    font-size: 2.25em;
    line-height: 1.2;
    border-bottom: 1px solid #eee;
}
h2 {
   padding-bottom: .3em;
    font-size: 1.75em;
    line-height: 1.225;
    border-bottom: 1px solid #eee;
}
h3 {
    font-size: 1.5em;
    line-height: 1.43;
}
h4 {
    font-size: 1.25em;
}
h5 {
    font-size: 1em;
}
h6 {
   font-size: 1em;
    color: #777;
}
p,
blockquote,
ul,
ol,
dl,
table{
    margin: 0.8em 0;
}
li>ol,
li>ul {
    margin: 0 0;
}
hr {
    height: 2px;
    padding: 0;
    margin: 16px 0;
    background-color: #e7e7e7;
    border: 0 none;
    overflow: hidden;
    box-sizing: content-box;
}

li p.first {
    display: inline-block;
}
ul,
ol {
    padding-left: 30px;
}
ul:first-child,
ol:first-child {
    margin-top: 0;
}
ul:last-child,
ol:last-child {
    margin-bottom: 0;
}
blockquote {
    border-left: 4px solid #dfe2e5;
    padding: 0 15px;
    color: #777777;
}
blockquote blockquote {
    padding-right: 0;
}
table {
    padding: 0;
    word-break: initial;
}
table tr {
    border-top: 1px solid #dfe2e5;
    margin: 0;
    padding: 0;
}
table tr:nth-child(2n),
thead {
    background-color: #f8f8f8;
}
table tr th {
    font-weight: bold;
    border: 1px solid #dfe2e5;
    border-bottom: 0;
    margin: 0;
    padding: 6px 13px;
}
table tr td {
    border: 1px solid #dfe2e5;
    margin: 0;
    padding: 6px 13px;
}
table tr th:first-child,
table tr td:first-child {
    margin-top: 0;
}
table tr th:last-child,
table tr td:last-child {
    margin-bottom: 0;
}

.CodeMirror-lines {
    padding-left: 4px;
}

.code-tooltip {
    box-shadow: 0 1px 1px 0 rgba(0,28,36,.3);
    border-top: 1px solid #eef2f2;
}

.md-fences,
code,
tt {
    border: 1px solid #e7eaed;
    background-color: #f8f8f8;
    border-radius: 3px;
    padding: 0;
    padding: 2px 4px 0px 4px;
    font-size: 0.9em;
}

code {
    background-color: #f3f4f4;
    padding: 0 2px 0 2px;
}

.md-fences {
    margin-bottom: 15px;
    margin-top: 15px;
    padding-top: 8px;
    padding-bottom: 6px;
}


.md-task-list-item > input {
  margin-left: -1.3em;
}

@media print {
    html {
        font-size: 13px;
    }
    table,
    pre {
        page-break-inside: avoid;
    }
    pre {
        word-wrap: break-word;
    }
}

.md-fences {
	background-color: #f8f8f8;
}
#write pre.md-meta-block {
	padding: 1rem;
    font-size: 85%;
    line-height: 1.45;
    background-color: #f7f7f7;
    border: 0;
    border-radius: 3px;
    color: #777777;
    margin-top: 0 !important;
}

.mathjax-block>.code-tooltip {
	bottom: .375rem;
}

.md-mathjax-midline {
    background: #fafafa;
}

#write>h3.md-focus:before{
	left: -1.5625rem;
	top: .375rem;
}
#write>h4.md-focus:before{
	left: -1.5625rem;
	top: .285714286rem;
}
#write>h5.md-focus:before{
	left: -1.5625rem;
	top: .285714286rem;
}
#write>h6.md-focus:before{
	left: -1.5625rem;
	top: .285714286rem;
}
.md-image>.md-meta {
    /*border: 1px solid #ddd;*/
    border-radius: 3px;
    padding: 2px 0px 0px 4px;
    font-size: 0.9em;
    color: inherit;
}

.md-tag {
    color: #a7a7a7;
    opacity: 1;
}

.md-toc { 
    margin-top:20px;
    padding-bottom:20px;
}

.sidebar-tabs {
    border-bottom: none;
}

#typora-quick-open {
    border: 1px solid #ddd;
    background-color: #f8f8f8;
}

#typora-quick-open-item {
    background-color: #FAFAFA;
    border-color: #FEFEFE #e5e5e5 #e5e5e5 #eee;
    border-style: solid;
    border-width: 1px;
}

/** focus mode */
.on-focus-mode blockquote {
    border-left-color: rgba(85, 85, 85, 0.12);
}

header, .context-menu, .megamenu-content, footer{
    font-family: "Segoe UI", "Arial", sans-serif;
}

.file-node-content:hover .file-node-icon,
.file-node-content:hover .file-node-open-state{
    visibility: visible;
}

.mac-seamless-mode #typora-sidebar {
    background-color: #fafafa;
    background-color: var(--side-bar-bg-color);
}

.md-lang {
    color: #b4654d;
}

.html-for-mac .context-menu {
    --item-hover-bg-color: #E6F0FE;
}

#md-notification .btn {
    border: 0;
}

.dropdown-menu .divider {
    border-color: #e5e5e5;
}

.ty-preferences .window-content {
    background-color: #fafafa;
}

.ty-preferences .nav-group-item.active {
    color: white;
    background: #999;
}

 .typora-export li, .typora-export p, .typora-export,  .footnote-line {white-space: normal;} 
</style>
</head>
<body class='typora-export os-windows' >
<div  id='write'  class = ''><h1><a name="celebrationnight" class="md-header-anchor"></a><span>CelebrationNight</span></h1><p><strong><span>This project was released under GPL-3.0 License.</span></strong></p><h2><a name="scripts-for-transcriptome-analysis" class="md-header-anchor"></a><span>Scripts for transcriptome analysis.</span></h2><ul><li><span>This project include some experiment coding for testing purpose.</span></li><li><span>Use it at your own risk.</span></li><li><span>I will try my best to complete and polish this documentation.</span></li></ul><h2><a name="type-of-process" class="md-header-anchor"></a><span>Type of process</span></h2><h3><a name="cofril" class="md-header-anchor"></a><span>CoFRIL</span></h3><ul><li><strong><span>Co</span></strong><span>nfig </span><strong><span>F</span></strong><span>irst and </span><strong><span>R</span></strong><span>un </span><strong><span>I</span></strong><span>t </span><strong><span>L</span></strong><span>ater</span></li></ul><figure><table><thead><tr><th><strong><span>Definition</span></strong></th><th><strong><span>Naming</span></strong></th></tr></thead><tbody><tr><td><span>1. Pack process code into python class (</span><strong><span>.py</span></strong><span> file with &quot;lib&quot; prefix)</span></td><td>&nbsp;</td></tr><tr><td><span>2. Run </span><span>*</span><span>-setConfig.py first to configure dependent variables (Initialization)</span></td><td><span>*</span><span>-setConfig.py</span></td></tr><tr><td>&nbsp;</td><td><strong><span>template</span></strong><span>:  &quot;</span><span>*</span><span>-configExample.py&quot;</span></td></tr><tr><td><span>3. Run related python class in </span><span>*</span><span>-run.py</span></td><td><span>*</span><span>-run.py</span></td></tr><tr><td>&nbsp;</td><td><strong><span>template</span></strong><span>:  &quot;</span><span>*</span><span>-exampleRun.py&quot;</span></td></tr></tbody></table></figure><h3><a name="rera" class="md-header-anchor"></a><span>ReRa</span></h3><ul><li><strong><span>Re</span></strong><span>lay </span><strong><span>Ra</span></strong><span>ce</span></li></ul><figure><table><thead><tr><th><strong><span>Definition</span></strong></th></tr></thead><tbody><tr><td><span>1. Use script language to control the workflow</span></td></tr><tr><td><span>2. Use executable binary or script language or both to process information</span></td></tr></tbody></table></figure><h3><a name="rudi" class="md-header-anchor"></a><span>RuDi</span></h3><ul><li><strong><span>Ru</span></strong><span>n </span><strong><span>Di</span></strong><span>rectly</span></li></ul><figure><table><thead><tr><th><strong><span>Definition</span></strong></th></tr></thead><tbody><tr><td><span>1. Declare sample-dependent variables on the beginning of script</span></td></tr><tr><td><span>2. Derive downstream variables</span></td></tr><tr><td><span>3. Run code directly without class</span></td></tr></tbody></table></figure><h2><a name="configuration" class="md-header-anchor"></a><span>Configuration</span></h2><h3><a name="libconfig" class="md-header-anchor"></a><span>libConfig</span></h3><p><strong><span>For general-purpose  storage of configuration</span></strong></p><ul><li><em><span>Coming soon</span></em></li></ul><h3><a name="cf-branch" class="md-header-anchor"></a><span>cf-Branch</span></h3><p><strong><span>For experiment design</span></strong></p><ul><li><em><span>Coming soon</span></em></li></ul><h3><a name="cf-command" class="md-header-anchor"></a><span>cf-Command</span></h3><p><strong><span>For version management of executable</span></strong></p><ul><li><em><span>Coming soon</span></em></li></ul><h3><a name="cf-parameter" class="md-header-anchor"></a><span>cf-Parameter</span></h3><p><strong><span>For adjustment</span></strong></p><ul><li><em><span>Coming soon</span></em></li></ul><h2><a name="database-conversion-preparation" class="md-header-anchor"></a><span>Database conversion (Preparation)</span></h2><h3><a name="extract-information-from-genome-annotation" class="md-header-anchor"></a><span>Extract information from Genome annotation</span></h3><ul><li><em><span>Coming soon</span></em></li></ul><h3><a name="reconstruct-information-from-functional-annotation-database" class="md-header-anchor"></a><span>Reconstruct information from functional annotation database</span></h3><ul><li><em><span>Coming soon</span></em></li></ul><h2><a name="processing-stage-general-usage" class="md-header-anchor"></a><span>Processing Stage (General usage)</span></h2><h3><a name="stage-01---fastq-quality-report" class="md-header-anchor"></a><span>Stage 01 - FASTQ Quality Report</span></h3><figure><table><thead><tr><th><span>List</span></th><th><span>Detail</span></th></tr></thead><tbody><tr><td><span>Codename</span></td><td><strong><span>01-fq-fastqc</span></strong></td></tr><tr><td><span>Usage</span></td><td><span>Quality reports</span></td></tr><tr><td><span>Type</span></td><td><span>ReRa</span></td></tr><tr><td><span>Binary</span></td><td><span>FastQC </span><sup><a href='#fastqc-ref'><span>1</span></a></sup></td></tr><tr><td><span>Language</span></td><td><span>Shell script &amp; Python 3</span></td></tr><tr><td><span>Input</span></td><td><span>NGS reads, FASTQ format</span></td></tr><tr><td><span>Output</span></td><td><span>HTML reports</span></td></tr></tbody></table></figure><ul><li><p><strong><span>Command:</span></strong><span> </span></p><ul><li><code>bin/FastQC/fastqc -f fastq -o [largeData/04-hisat2/species/speciesDatabase-trimQ30/report]</code></li></ul></li></ul><h3><a name="stage-02---hisat2-index" class="md-header-anchor"></a><span>Stage 02 - HISAT2 index</span></h3><figure><table><thead><tr><th><span>List</span></th><th><span>Detail</span></th></tr></thead><tbody><tr><td><span>Codename</span></td><td><strong><span>02-hisat2-index</span></strong></td></tr><tr><td><span>Usage</span></td><td><span>Build HISAT2 index</span></td></tr><tr><td><span>Type</span></td><td><span>CoFRIL</span></td></tr><tr><td><span>Class</span></td><td><span>libHISAT.indexer() </span><sup><a href='#indexer'><span>indexer</span></a></sup></td></tr><tr><td><span>Binary</span></td><td><span>hisat2-build from HISAT2 </span><sup><a href='#tuxado-ref'><span>2</span></a></sup></td></tr><tr><td><span>Input</span></td><td><span>Genome sequences, FASTA format</span></td></tr><tr><td><span>Output</span></td><td><span>HISAT2 index of genome</span></td></tr></tbody></table></figure><h3><a name="stage-03---trim" class="md-header-anchor"></a><span>Stage 03 - Trim</span></h3><figure><table><thead><tr><th><span>List</span></th><th><span>Detail</span></th></tr></thead><tbody><tr><td><span>Codename</span></td><td><strong><span>03-trim</span></strong></td></tr><tr><td><span>Usage</span></td><td><span>Trim FASTQ files</span></td></tr><tr><td><span>Type</span></td><td><span>CoFRIL</span></td></tr><tr><td><span>Class</span></td><td><span>libTrim.trimmer() </span><sup><a href='#indexer'><span>indexer</span></a></sup></td></tr><tr><td><span>Binary</span></td><td><span>Trimmomatic </span><sup><a href='#trim-ref'><span>3</span></a></sup></td></tr><tr><td><span>Input</span></td><td><span>raw NGS reads, FASTQ format</span></td></tr><tr><td><span>Output</span></td><td><span>Trimmed reads, FASTQ format</span></td></tr></tbody></table></figure><h3><a name="stage-04---hisat2" class="md-header-anchor"></a><span>Stage 04 - HISAT2</span></h3><figure><table><thead><tr><th><span>List</span></th><th><span>Detail</span></th></tr></thead><tbody><tr><td><span>Codename</span></td><td><strong><span>04-hisat2</span></strong></td></tr><tr><td><span>Usage</span></td><td><span>Alignment and mapping</span></td></tr><tr><td><span>Type</span></td><td><span>CoFRIL</span></td></tr><tr><td><span>Class</span></td><td><span>libHISAT.aligner() </span><sup><a href='#aligner'><span>aligner</span></a></sup></td></tr><tr><td><span>Binary</span></td><td><span>hisat2 from HISAT2 </span><sup><a href='#tuxado-ref'><span>2</span></a></sup></td></tr><tr><td>&nbsp;</td><td><span>samtools from SAMtools </span><sup><a href='#samtool-ref'><span>4</span></a></sup></td></tr><tr><td><span>Input</span></td><td><span>HISAT2 index of genome </span><strong><span>(02-hisat2-index)</span></strong></td></tr><tr><td>&nbsp;</td><td><span>Trimmed reads, FASTQ format </span><strong><span>(03-trim)</span></strong></td></tr><tr><td><span>Output</span></td><td><span>Alignments, BAM format</span></td></tr></tbody></table></figure><h3><a name="stage-05---gffread" class="md-header-anchor"></a><span>Stage 05 - gffRead</span></h3><figure><table><thead><tr><th><span>List</span></th><th><span>Detail</span></th></tr></thead><tbody><tr><td><span>Codename</span></td><td><strong><span>05-gr-gffRead</span></strong></td></tr><tr><td><span>Usage</span></td><td><span>Convert genome into transcriptome</span></td></tr><tr><td><span>Type</span></td><td><span>CoFRIL</span></td></tr><tr><td><span>Class</span></td><td><span>libCuffdiff.converter() </span><sup><a href='#converter'><span>converter</span></a></sup></td></tr><tr><td><span>Binary</span></td><td><span>gffread </span><sup><a href='#gff-ref'><span>5</span></a></sup></td></tr><tr><td><span>Input</span></td><td><span>Genome annotation, GFF3 format</span></td></tr><tr><td><span>Output</span></td><td><span>Transcriptome annotation, GTF2 format</span></td></tr></tbody></table></figure><h3><a name="stage-06---transcripts-extraction" class="md-header-anchor"></a><span>Stage 06 - Transcripts extraction</span></h3><figure><table><thead><tr><th><span>List</span></th><th><span>Detail</span></th></tr></thead><tbody><tr><td><span>Codename</span></td><td><strong><span>06-fn</span></strong></td></tr><tr><td><span>Usage</span></td><td><span>Transcriptome extractor</span></td></tr><tr><td><span>Type</span></td><td><span>ReRa</span></td></tr><tr><td><span>Binary</span></td><td><span>gffread </span><sup><a href='#gff-ref'><span>5</span></a></sup></td></tr><tr><td><span>Language</span></td><td><span>Shell script &amp; Python 3</span></td></tr><tr><td><span>Input</span></td><td><span>Genome sequences, FASTA format</span></td></tr><tr><td>&nbsp;</td><td><span>Transcriptome annotation, GTF2 format </span><strong><span>(05-gr-gffRead)</span></strong></td></tr><tr><td><span>Output</span></td><td><span>Transcripts, FASTA format</span></td></tr><tr><td>&nbsp;</td><td><span>Transcripts table, TSV format</span></td></tr></tbody></table></figure><ul><li><p><strong><span>Command:</span></strong><span> </span></p><ul><li><code>bin/cufflinks/gffread\</code>
<code>-g userData/dbgs-GenomeSequence/speciesDatabase/speciesDatabase.fn\</code>
<code>-w userData/06-gr-exportTranscript/speciesTreatment/speciesDatabase-trimQ30-transcript.fn\</code>
<code>userData/05-gr-transcriptomeConstruction/speciesTreatment/speciesDatabase-trimQ30-final.gtf</code></li></ul></li></ul><h3><a name="stage-07---transcript-abundances-estimation" class="md-header-anchor"></a><span>Stage 07 - Transcript abundances estimation</span></h3><figure><table><thead><tr><th><span>List</span></th><th><span>Detail</span></th></tr></thead><tbody><tr><td><span>Codename</span></td><td><strong><span>07-cd-CuffDiff</span></strong></td></tr><tr><td><span>Usage</span></td><td><span>Estimate transcript abundances</span></td></tr><tr><td><span>Type</span></td><td><span>CoFRIL</span></td></tr><tr><td><span>Class</span></td><td><span>libCuffdiff.differ() </span><sup><a href='#differ'><span>differ</span></a></sup></td></tr><tr><td><span>Binary</span></td><td><span>cuffdiff from Cufflinks </span><sup><a href='#cuff-ref'><span>6</span></a></sup></td></tr><tr><td><span>Input</span></td><td><span>Alignments, BAM format </span><strong><span>(04-hisat2)</span></strong></td></tr><tr><td>&nbsp;</td><td><span>Transcriptome annotation, GTF2 format </span><strong><span>(05-gr-gffRead)</span></strong></td></tr><tr><td><span>Output</span></td><td><span>Abundances/Expression profile, DIFF format (TSV format)</span></td></tr></tbody></table></figure><h3><a name="stage-08---homologous-annotation" class="md-header-anchor"></a><span>Stage 08 - Homologous annotation</span></h3><figure><table><thead><tr><th><span>List</span></th><th><span>Detail</span></th></tr></thead><tbody><tr><td><span>Codename</span></td><td><strong><span>08-an</span></strong></td></tr><tr><td><span>Usage</span></td><td><span>an1. Extract info of transcript-gene relation</span></td></tr><tr><td>&nbsp;</td><td><span>an2. Link transcript ID with gene ID and homologous ID</span></td></tr><tr><td>&nbsp;</td><td><span>(cont.) Further annotate with functional annotation databases</span></td></tr><tr><td><span>Type</span></td><td><span>RuDi</span></td></tr><tr><td><span>Language</span></td><td><span>Python 3</span></td></tr><tr><td><span>Input</span></td><td><span>Abundances/Expression profile, DIFF format (TSV format) </span><strong><span>(07-cd-CuffDiff)</span></strong></td></tr><tr><td>&nbsp;</td><td><span>Genome annotation, JSON format </span><strong><span>(dbga-GenomeAnnotation)</span></strong></td></tr><tr><td>&nbsp;</td><td><span>Homolog , JSON format </span><strong><span>(dbga-GenomeAnnotation)</span></strong></td></tr><tr><td>&nbsp;</td><td><span>GO Terms, JSON format </span><strong><span>(dbgo-GOdatabase)</span></strong></td></tr><tr><td>&nbsp;</td><td><span>KEGG pathways, JSON format </span><strong><span>(dbkg-KEGG-hirTree)</span></strong></td></tr><tr><td>&nbsp;</td><td><span>Other databases...</span></td></tr><tr><td><span>Output</span></td><td><span>Homologous annotations, JSONs format (various files)</span></td></tr></tbody></table></figure><h3><a name="stage-09---transcriptome-information-summarizer" class="md-header-anchor"></a><span>Stage 09 - Transcriptome information summarizer</span></h3><figure><table><thead><tr><th><span>List</span></th><th><span>Detail</span></th></tr></thead><tbody><tr><td><span>Codename</span></td><td><strong><span>09-cd</span></strong></td></tr><tr><td><span>Usage</span></td><td><span>cd1. Convert the results of cuffdiff into SQLite3 form</span></td></tr><tr><td>&nbsp;</td><td><span>cd2. Annotate and seperate information into...</span></td></tr><tr><td>&nbsp;</td><td><span>(cont.) sample-orientation Expression tables</span></td></tr><tr><td><span>Type</span></td><td><span>RuDi</span></td></tr><tr><td><span>Language</span></td><td><span>Python 3</span></td></tr><tr><td><span>Input</span></td><td><span>Abundances/Expression profile, DIFF format (TSV format) </span><strong><span>(07-cd-CuffDiff)</span></strong></td></tr><tr><td>&nbsp;</td><td><span>Homologous annotations, JSONs format </span><strong><span>(08-an)</span></strong></td></tr><tr><td><span>Output</span></td><td><span>Annotated abundances/expression profile, SQLite3 format</span></td></tr><tr><td>&nbsp;</td><td><span>Annotated abundances/expression profile, TSV format</span></td></tr></tbody></table></figure><h3><a name="stage-10---differential-expression-analysis-havent-release" class="md-header-anchor"></a><span>Stage 10 - Differential Expression Analysis (Haven&#39;t release)</span></h3><figure><table><thead><tr><th><span>List</span></th><th><span>Detail</span></th></tr></thead><tbody><tr><td><span>Codename</span></td><td><strong><span>10-grouping</span></strong></td></tr><tr><td><span>Usage</span></td><td><span>Group and calculate the count of Differential Expressed Genes (DEGs) and Specifically Expressed Genes (SEGs)</span></td></tr><tr><td><span>Type</span></td><td><span>RuDi</span></td></tr><tr><td><span>Language</span></td><td><span>Python 3</span></td></tr><tr><td><span>Input</span></td><td><span>Annotated abundances/expression profile, TSV format </span><strong><span>(09-cd)</span></strong></td></tr><tr><td>&nbsp;</td><td><span>Homologous annotations, JSONs format </span><strong><span>(08-an)</span></strong></td></tr><tr><td><span>Output</span></td><td><span>DEA result files, JSON format</span></td></tr><tr><td>&nbsp;</td><td><span>DEA result files, TSV format</span></td></tr><tr><td>&nbsp;</td><td><span>DEA count record, LOG format (TXT format)</span></td></tr></tbody></table></figure><h3><a name="stage-11---gene-set-enrichment-analysis-havent-release" class="md-header-anchor"></a><span>Stage 11 - Gene set enrichment analysis (Haven&#39;t release)</span></h3><figure><table><thead><tr><th><span>List</span></th><th><span>Detail</span></th></tr></thead><tbody><tr><td><span>Codename</span></td><td><strong><span>11-ea-enrichAnaly</span></strong></td></tr><tr><td><span>Usage</span></td><td><span>Compare and calculate the ratio of count of DEGs or SEGs</span></td></tr><tr><td><span>Type</span></td><td><span>RuDi</span></td></tr><tr><td><span>Language</span></td><td><span>Python 3</span></td></tr><tr><td><span>Input</span></td><td><span>DEA result files, JSON format </span><strong><span>(10-grouping)</span></strong></td></tr><tr><td>&nbsp;</td><td><span>Annotated abundances/expression profile, SQLite3 format </span><strong><span>(09-cd)</span></strong></td></tr><tr><td>&nbsp;</td><td><span>Homologous annotations, JSONs format </span><strong><span>(08-an)</span></strong></td></tr><tr><td><span>Output</span></td><td><span>GSEA result files, JSON format</span></td></tr><tr><td>&nbsp;</td><td><span>GSEA result files, TSV format</span></td></tr></tbody></table></figure><h3><a name="stage-12---fishers-exact-test-and-visualization-havent-release" class="md-header-anchor"></a><span>Stage 12 - Fisher&#39;s Exact Test and visualization (Haven&#39;t release)</span></h3><figure><table><thead><tr><th><span>List</span></th><th><span>Detail</span></th></tr></thead><tbody><tr><td><span>Codename</span></td><td><strong><span>12-ft</span></strong></td></tr><tr><td><span>Usage</span></td><td><span>ft1. Do Fisher&#39;s Exact Test</span></td></tr><tr><td>&nbsp;</td><td><span>ft2. Visualization</span></td></tr><tr><td><span>Type</span></td><td><span>RuDi</span></td></tr><tr><td><span>Language</span></td><td><span>Python 3</span></td></tr><tr><td><span>Input</span></td><td><span>GSEA result files, JSON format </span><strong><span>(11-ea-enrichAnaly)</span></strong></td></tr><tr><td>&nbsp;</td><td><span>Homologous annotations, JSONs format </span><strong><span>(08-an)</span></strong></td></tr><tr><td><span>Output</span></td><td><span>FET result files, PNG format</span></td></tr><tr><td>&nbsp;</td><td><span>FET result files, SVG format</span></td></tr><tr><td>&nbsp;</td><td><span>FET result files, TSV format</span></td></tr></tbody></table></figure><h2><a name="python-library-module--classes" class="md-header-anchor"></a><span>Python Library (Module &amp; Classes)</span></h2><h3><a name="libhisat" class="md-header-anchor"></a><span>libHISAT</span></h3><ul><li><p><span>Original Command:</span></p><ul><li><p><strong><a name="indexer"><span>libHISAT.indexer()</span></a><span> </span></strong></p><ul><li><code>hisat2-build \</code><span> </span><strong><span>#Building HISAT2 Index</span></strong>
<code>-p [THREAD] &lt;Path and Name of GENOME File&gt; \</code><span> </span>
<code>&lt; prefix of HISAT2-build genome index (path+header)&gt;</code></li></ul></li><li><p><strong><a name="aligner"><span>libHISAT.aligner()</span></a><span> </span></strong></p><ul><li><code>hisat2 \</code><span> </span><strong><span>#Aligning and mapping</span></strong>
<code>-q [--dta/--dta-cufflinks] --phred &lt;phred&gt; -p &lt;thread&gt; \</code>
<code>-x &lt;prefix of HISAT2-build genome index&gt; \</code>
<code>-1 &lt;forward fastq files of samples&gt; \</code>
<code>-2 &lt;reverse fastq files of samples&gt; \</code>
<code>-S &lt;output SAM files&gt; \</code></li><li><code>samtools view -o &lt;out.bam&gt; -Su &lt;in.sam&gt;</code><span> </span><strong><span>#Convert SAM to BAM</span></strong></li><li><code>samtools sort -o &lt;out-sorted.bam&gt; &lt;in.bam&gt;</code><span> </span><strong><span>#Sorting BAM for decreasing file size</span></strong></li></ul></li></ul></li></ul><h3><a name="libtrim" class="md-header-anchor"></a><span>libTrim</span></h3><ul><li><p><span>Original Command:</span></p><ul><li><p><strong><a name="trimmer"><span>libTrim.trimmer()</span></a></strong></p><ul><li><code>java -jar &lt;bin&gt;/trimmomatic-0.35.jar PE \</code><span> </span><strong><span>#Pair-End</span></strong>
<code>-phred33 -threads &lt;threads&gt; \</code>
<code>input_forward.fq.gz input_reverse.fq.gz \</code>
<code>output_forward_paired.fq.gz output_forward_unpaired.fq.gz \</code>
<code>output_reverse_paired.fq.gz output_reverse_unpaired.fq.gz \</code>
<code>&lt;ILLUMINACLIP&gt; &lt;LEADING&gt; \</code>
<code>&lt;TRAILING&gt; &lt;SLIDINGWINDOW&gt; &lt;MINLEN&gt;</code></li><li><code>java -jar &lt;bin&gt;/trimmomatic-0.35.jar SE \</code><span> </span><strong><span>#Single-End</span></strong>
<code>-phred33 -threads &lt;threads&gt; \</code>
<code>input.fq.gz output.fq.gz \</code>
<code>&lt;ILLUMINACLIP&gt; &lt;LEADING&gt; \</code>
<code>&lt;TRAILING&gt; &lt;SLIDINGWINDOW&gt; &lt;MINLEN&gt;</code></li></ul></li></ul></li></ul><h3><a name="libcuffdiff" class="md-header-anchor"></a><span>libCuffdiff</span></h3><ul><li><p><span>Original Command:</span></p><ul><li><p><strong><a name="differ"><span>libCuffdiff.differ()</span></a></strong></p><ul><li><code>cuffdiff \``</code><span>-p </span><int><span>  -o </span><string><span> </span><span>`</span>
<code>-L &lt;label1,label2,…,labelN&gt; &lt;transcripts.gtf&gt; \</code>
<code>[[sample1_replicate1.sam,…] …… […,sampleN_replicateM.sam]]</code></li></ul></li><li><p><strong><a name="converter"><span>libCuffdiff.converter()</span></a></strong></p><ul><li><code>bin/cufflinks/gffread &lt;inputFile&gt; -T -o &lt;outputFile&gt;</code></li></ul></li></ul></li></ul><h2><a name="references" class="md-header-anchor"></a><span>References</span></h2><ol start='' ><li><p><span>[</span><a name="fastqc-ref"><span>FastQC</span></a><span>] </span><a href='https://www.bioinformatics.babraham.ac.uk/projects/fastqc/' target='_blank' class='url'>https://www.bioinformatics.babraham.ac.uk/projects/fastqc/</a></p><ul><li><strong><span>Andrews,  S.</span></strong><span> (2018).  FastQC:  a  quality  control  tool  for  high  throughput  sequence  data (Babraham Bioinformatics, Babraham Institute, Cambridge, United Kingdom). </span></li></ul></li><li><p><span>[</span><a name="tuxado-ref"><span>HISAT2</span></a><span>] </span><a href='https://doi.org/10.1038/nprot.2016.095' target='_blank' class='url'>https://doi.org/10.1038/nprot.2016.095</a><span> (Article)</span></p><ul><li><strong><span>Pertea, M., Kim, D., Pertea, G.M., Leek, J.T., and Salzberg, S.L.</span></strong><span> (2016). Transcript-level  expression  analysis  of  RNA-seq  experiments  with  HISAT,  StringTie  and Ballgown. Nat Protoc </span><strong><span>11,</span></strong><span> 1650-1667.</span></li><li><a href='https://ccb.jhu.edu/software/hisat2/manual.shtml' target='_blank' class='url'>https://ccb.jhu.edu/software/hisat2/manual.shtml</a><span> (Documentation &amp; Binary)</span></li></ul></li><li><p><span>[</span><a name="trim-ref"><span>Trimmomatic</span></a><span>] </span><a href='https://doi.org/10.1093/bioinformatics/btu170' target='_blank' class='url'>https://doi.org/10.1093/bioinformatics/btu170</a><span> (Article)</span></p><ul><li><strong><span>Bolger, A.M., Lohse, M., and Usadel, B.</span></strong><span> (2014). Trimmomatic: a flexible trimmer for Illumina sequence data. Bioinformatics </span><strong><span>30,</span></strong><span> 2114-2120.</span></li><li><a href='http://www.usadellab.org/cms/?page=trimmomatic' target='_blank' class='url'>http://www.usadellab.org/cms/?page=trimmomatic</a><span> (Documentation &amp; Binary)</span></li></ul></li><li><p><span>[</span><a name="samtool-ref"><span>SAMtools</span></a><span>] </span><a href='https://www.htslib.org/doc/samtools.html' target='_blank' class='url'>https://www.htslib.org/doc/samtools.html</a></p></li><li><p><span>[</span><a name="gff-ref"><span>gffread</span></a><span>] </span><a href='https://ccb.jhu.edu/software/stringtie/gff.shtml' target='_blank' class='url'>https://ccb.jhu.edu/software/stringtie/gff.shtml</a></p></li><li><p><span>[</span><a name="cuff-ref"><span>Cuffdiff</span></a><span>]  </span><a href='https://doi.org/10.1038/nbt.2450' target='_blank' class='url'>https://doi.org/10.1038/nbt.2450</a><span> (Article)</span></p><ul><li><strong><span>Trapnell, C., Hendrickson, D.G., Sauvageau, M., Goff, L., Rinn, J.L., and Pachter,</span>
<span>L.</span></strong><span> (2013). Differential analysis of gene regulation at transcript resolution with</span>
<span>RNA-seq. Nat Biotechnol </span><strong><span>31,</span></strong><span> 46-53.</span></li><li><a href='https://cole-trapnell-lab.github.io/cufflinks/manual/' target='_blank' class='url'>https://cole-trapnell-lab.github.io/cufflinks/manual/</a><span> (Documentation)</span></li><li><a href='https://cole-trapnell-lab.github.io/cufflinks/install/' target='_blank' class='url'>https://cole-trapnell-lab.github.io/cufflinks/install/</a><span> (Binary)</span></li></ul></li></ol><h2><a name="processing-stage-distinctive-usage" class="md-header-anchor"></a><span>Processing Stage (Distinctive usage)</span></h2><p><strong><span>The scripts under this catalogue may lose function as their development didn&#39;t stick to the current coding style.</span></strong></p><h3><a name="stage-04---hisat2-summariser" class="md-header-anchor"></a><span>Stage 04 - HISAT2 Summariser</span></h3><figure><table><thead><tr><th><span>List</span></th><th><span>Detail</span></th></tr></thead><tbody><tr><td><span>Codename</span></td><td><strong><span>04-hs</span></strong></td></tr><tr><td><span>Usage</span></td><td><span>Analyse HISAT2 result</span></td></tr><tr><td><span>Type</span></td><td><span>CoFRIL</span></td></tr><tr><td><span>Class</span></td><td><span>libHISAT.summariser()</span></td></tr><tr><td><span>Binary</span></td><td><span>samtools from SAMtools</span></td></tr><tr><td><span>Input</span></td><td><strong><span>04-hisat2</span></strong></td></tr></tbody></table></figure><h3><a name="stage-07---comparing-genomic-annotation" class="md-header-anchor"></a><span>Stage 07 - Comparing Genomic Annotation</span></h3><figure><table><thead><tr><th><span>List</span></th><th><span>Detail</span></th></tr></thead><tbody><tr><td><span>Codename</span></td><td><strong><span>07-cg</span></strong></td></tr><tr><td><span>Usage</span></td><td><span>Get information of isoform under each gene model</span></td></tr><tr><td><span>Type</span></td><td><span>RuDi</span></td></tr><tr><td><span>Input</span></td><td><strong><span>07-st</span></strong></td></tr></tbody></table></figure></div>
</body>
</html>